tests/main/views/test_drafts.py

```diff
@@ -56,6 +62,455 @@ def service_count(self):
+            Service.supplier_id == 1,
+            Service.framework_id == 1,  # G-Cloud 6
+            Service.lot_id == 2,
+        ).all()
+
+        res = self.post_to_copy_published_from_framework()
+        assert res.status_code == 201
+
+        data = json.loads(res.get_data(as_text=True))
+        serialized_drafts = data['services']
+
+        db_drafts = DraftService.query.filter(
+            DraftService.supplier_id == 1,
+            DraftService.framework_id == 4,  # G-Cloud 7
+            DraftService.lot_id == 2,
+        ).all()
+
+        assert len(serialized_drafts) == 5
+        assert len(db_drafts) == 5
+        assert all(service['status'] == 'not-submitted' for service in serialized_drafts)
+        assert all(service['frameworkSlug'] == 'g-cloud-7' for service in serialized_drafts)
+        assert all(not service.get('serviceId') for service in serialized_drafts)
+        for service, serialized_draft, db_draft in zip(services, serialized_drafts, db_drafts):
+            assert service.data['serviceName'] == serialized_draft['serviceName'] == db_draft.data['serviceName']
+
+    def test_should_400_if_no_source_framework_slug(self):
+        res = self.post_to_copy_published_from_framework(source_framework_slug=None)
+
+        assert res.status_code == 400
+        assert "Required data missing: 'source_framework_slug'" in res.get_data(as_text=True)
+
+    def test_should_400_if_no_questions_to_copy(self):
+        res = self.post_to_copy_published_from_framework(questions_to_copy=None)
+
+        assert res.status_code == 400
+        assert "Required data missing: 'questions_to_copy'" in json.loads(res.get_data(as_text=True))['error']
+
+    def test_should_404_if_framework_does_not_exist(self):
+        res = self.post_to_copy_published_from_framework(framework_slug='z-cloud')
+
+        assert res.status_code == 404
+
+    @pytest.mark.parametrize('framework_status', Framework.STATUSES)
+    def test_should_400_if_target_framework_not_open(self, framework_status):
+        if framework_status == 'open':
+            return
+
+        self.set_framework_status('g-cloud-7', framework_status)
+
+        res = self.post_to_copy_published_from_framework()
+        assert res.status_code == 400
+
+        assert 'Target framework is not open' in res.get_data(as_text=True)
+
+    def test_should_only_copy_published_services_from_correct_framework_and_lot(self):
+        self.setup_dummy_services(3, supplier_id=1, lot_id=1, status='published')
+        self.setup_dummy_services(2, supplier_id=1, lot_id=1, status='disabled', start_id=3)
+        self.setup_dummy_services(4, supplier_id=1, lot_id=2, status='published', start_id=5)
+        self.setup_dummy_services(1, supplier_id=1, lot_id=2, status='disabled', start_id=9)
+        self.setup_dummy_services(2, supplier_id=1, lot_id=3, status='published', start_id=10)
+        self.setup_dummy_services(3, supplier_id=1, lot_id=3, status='enabled', start_id=12)
```

Comment:
Might be nice to test that only services from the same supplier are copied across. You could test that "for free" by creating a service here that's published on the right lot by a different supplier, but it might be clearer as a separate test. It would obviously be _completely broken_ if this condition wasn't met, but it's still good to explicitly test it.

Reply:
Have added the extra test